### DISCUSSION TITLE
Flex Vs PhysX Particle System - part 2

### DISCUSSION THREAD

**Discussion by trsh** [NVIDIA-Omniverse/PhysX#129] [Q&A]
Fallowing up on https://github.com/NVIDIA-Omniverse/PhysX/discussions/124, I see some of the important questions are still unanswered, like:

> Are all Flex features in PhysX Particle System? or same features less in PhysX Particle System?

So Flex was a full flagged particle PBD simulator very rich in examples. Going trough those one can realize, that its possible to simulate almost anything - fluids, cloth, soft bodies, granual sets, rigid bodies and even transitions between those (very interesting part), like in solids melting example. All of them were also able interact with each other. I think its like the holy idea of PBD, to simulate most of the physical phenomenon with single united mechanism.

In contrast the new system has only 4 snippet examples. And while lurking in code, it feels like something rather new, and not an adaption from Flex (maybe I am wrong). But here we talk about granual particles, fluids, rigids and cloth, and that's it. Not any more about soft/plastic bodies, etc.

So if a developer like me wants all the goodies I see in Flex, it becomes very puzzling if the new system can deliver it (besides the Cuda only restriction - thats clear). One could easily just stick with Flex, but that isn't smart either, as the support seems to be dropped for almost 2 years.

We are kindly asking Nvidia to shine a light a bit on this. Like what is the history of this new Particle System. Is it an adaption from Flex, or is it from starch? Does it even make sense to compare it with Flex and expect same feature set? Anything that would lift this cofunction and bring some more clarity. 

Also I created/copied a list with all original Flex examples (below). Maybe Nvidia can add comment for each: a) `Not possible` with new system; b) `Possible`; c) `Partly` (how?); d) `Never will be`; e) `Maybe in future` or similar. Maybe a migration guide can still be an option?

I know this could be annoying thing todo for nvidia devs (explaining and documenting, rather then moving forward), but I see it as a necessarily at this moment, and question like this will continue to pop out. There are more flex users, then it seems, because not all project where completed or/and made the way into being recognizable.

_P.S. please lets not involve other systems, like separate Soft Body tech that can interact with this one, but rather understand what is possible with PBD particles system alone_

```
// collision scenes
FrictionRamp("Friction Ramp")
FrictionMovingShape("Friction Moving Box", 0)
FrictionMovingShape("Friction Moving Sphere", 1)
FrictionMovingShape("Friction Moving Capsule", 2)
FrictionMovingShape("Friction Moving Mesh", 3)
ShapeCollision("Shape Collision")
ShapeChannels("Shape Channels")
TriangleCollision("Triangle Collision")
LocalSpaceFluid("Local Space Fluid")
LocalSpaceCloth("Local Space Cloth")
CCDFluid("World Space Fluid")

// cloth scenes
EnvironmentalCloth("Env Cloth Small", 6, 6, 40, 16)
EnvironmentalCloth("Env Cloth Large", 16, 32, 10, 3)
FlagCloth("Flag Cloth")
Inflatable("Inflatables")
ClothLayers("Cloth Layers")
SphereCloth("Sphere Cloth")
Tearing("Tearing")
Pasta("Pasta")

// game mesh scenes
GameMesh("Game Mesh Rigid", 0)
GameMesh("Game Mesh Particles", 1)
GameMesh("Game Mesh Fluid", 2)
GameMesh("Game Mesh Cloth", 3)
RigidDebris("Rigid Debris")

// viscous fluids
Viscosity("Viscosity Low", 0.5f)
Viscosity("Viscosity Med", 3.0f)
Viscosity("Viscosity High", 5.0f, 0.12f)
Adhesion("Adhesion")
GooGun("Goo Gun", true)

// regular fluids
Buoyancy("Buoyancy")
Melting("Melting")
SurfaceTension("Surface Tension Low", 0.0f)
SurfaceTension("Surface Tension Med", 10.0f)
SurfaceTension("Surface Tension High", 20.0f)
DamBreak("DamBreak  5cm", 0.05f)
DamBreak("DamBreak 10cm", 0.1f)
DamBreak("DamBreak 15cm", 0.15f)
RockPool("Rock Pool")
RayleighTaylor2D("Rayleigh Taylor 2D")

// misc feature scenes
TriggerVolume("Trigger Volume")
ForceField("Force Field")
InitialOverlap("Initial Overlap")

// rigid body scenes
RigidPile("Rigid2", 2)
RigidPile("Rigid4", 4)
RigidPile("Rigid8", 12)
BananaPile("Bananas")
LowDimensionalShapes("Low Dimensional Shapes")

// granular scenes
GranularPile("Granular Pile")

// coupling scenes
ParachutingBunnies("Parachuting Bunnies")
WaterBalloon("Water Balloons")
RigidFluidCoupling("Rigid Fluid Coupling")
FluidBlock("Fluid Block")
FluidClothCoupling("Fluid Cloth Coupling Water", false)
FluidClothCoupling("Fluid Cloth Coupling Goo", true)
BunnyBath("Bunny Bath Dam", true)
```

**Comment by sschirm** [comment]
Hi @trsh, 

The PhysX particle system is indeed not a straight port of Flex, and it will likely never become a 1:1 replacement. Flex went a long way being a general simulator using only particles which comes with great uniformity at the interface level and provides possibilities which other systems don't provide that easily like phase transitions. It also has disadvantages for very common use-cases. PhysX is trying to strike another balance.

> please lets not involve other systems, like separate Soft Body tech that can interact with this one, but rather understand what is possible with PBD particles system alone

I don't necessarily agree with this premise, as a lot of interactions and behaviors can be done better as a mixed simulation model all integrated in a unified solver as PhysX provides.

Making PhysX reach feature parity to Flex is currently not a priority for us.

Cheers, 
Simon

**Reply by trsh** [reply]
> Hi @trsh,
> 
> The PhysX particle system is indeed not a straight port of Flex, and it will likely never become a 1:1 replacement. Flex went a long way being a general simulator using only particles which comes with great uniformity at the interface level and provides possibilities which other systems don't provide that easily like phase transitions. It also has disadvantages for very common use-cases. PhysX is trying to strike another balance.

So if I want Flex like `uniformity` and `phase transitions`, its a dead end with new Particle System?

> I don't necessarily agree with this premise, as a lot of interactions and behaviors can be done better as a mixed simulation model all integrated in a unified solver as PhysX provides.
> 
> Making PhysX reach feature parity to Flex is currently not a priority for us.
> 
Yes but it kills your mentioned `uniformity` and `phase transitions`. My project strongly relies on a unified solution. In this case I would be switching back and forth, like when hard object becomes soft and beyond.
